src/audio/mfcc/mfcc.c

```diff
@@ -0,0 +1,277 @@
+
+	comp_info(mod->dev, "mfcc_free()");
+	comp_data_blob_handler_free(cd->model_handler);
+	mfcc_free_buffers(cd);
+	rfree(cd);
+	return 0;
+}
+
+static int mfcc_get_config(struct processing_module *mod,
+			   uint32_t config_id, uint32_t *data_offset_size,
+			   uint8_t *fragment, size_t fragment_size)
+{
+	struct sof_ipc_ctrl_data *cdata = (struct sof_ipc_ctrl_data *)fragment;
+	struct mfcc_comp_data *cd = module_get_private_data(mod);
+
+	comp_info(mod->dev, "mfcc_get_config()");
+
+	return comp_data_blob_get_cmd(cd->model_handler, cdata, fragment_size);
+}
+
+static int mfcc_set_config(struct processing_module *mod, uint32_t config_id,
+			   enum module_cfg_fragment_position pos, uint32_t data_offset_size,
+			   const uint8_t *fragment, size_t fragment_size, uint8_t *response,
+			   size_t response_size)
+{
+	struct mfcc_comp_data *cd = module_get_private_data(mod);
+
+	comp_info(mod->dev, "mfcc_set_config()");
+
+	return comp_data_blob_set(cd->model_handler, pos, data_offset_size,
+				  fragment, fragment_size);
+}
+
+static int mfcc_process(struct processing_module *mod,
+			struct input_stream_buffer *input_buffers, int num_input_buffers,
+			struct output_stream_buffer *output_buffers, int num_output_buffers)
+{
+	struct mfcc_comp_data *cd = module_get_private_data(mod);
+	struct audio_stream __sparse_cache *source = input_buffers->data;
+	struct audio_stream __sparse_cache *sink = output_buffers->data;
+	int frames = input_buffers->size;
+
+	comp_dbg(mod->dev, "mfcc_process(), start");
+
+	frames = MIN(frames, cd->max_frames);
+	cd->mfcc_func(mod, input_buffers, output_buffers, frames);
+
+	/* TODO: use module_update_buffer_position() from #6194 */
+	input_buffers->consumed += audio_stream_frame_bytes(source) * frames;
+	output_buffers->size += audio_stream_frame_bytes(sink) * frames;
+	comp_dbg(mod->dev, "mfcc_process(), done");
+	return 0;
+}
+
+static void mfcc_set_alignment(struct audio_stream *source, struct audio_stream *sink)
+{
+	const uint32_t byte_align = 1;
+	const uint32_t frame_align_req = 1;
+
+	audio_stream_init_alignment_constants(byte_align, frame_align_req, source);
+	audio_stream_init_alignment_constants(byte_align, frame_align_req, sink);
+}
+
+static int mfcc_prepare(struct processing_module *mod)
+{
+	struct mfcc_comp_data *cd = module_get_private_data(mod);
+	struct comp_buffer *sourceb;
+	struct comp_buffer *sinkb;
+	struct comp_buffer __sparse_cache *source_c;
+	struct comp_buffer __sparse_cache *sink_c;
+	struct comp_dev *dev = mod->dev;
+	enum sof_ipc_frame source_format;
+	enum sof_ipc_frame sink_format;
+	uint32_t sink_period_bytes;
+	int ret;
+
+	comp_info(dev, "mfcc_prepare()");
+
+	/* MFCC component will only ever have 1 source and 1 sink buffer */
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	source_c = buffer_acquire(sourceb);
+	sink_c = buffer_acquire(sinkb);
```

Comment:
you have to release the buffers - both on the error and the success paths

Reply:
Oops, I think I had other violations in mfcc_prepare() for sourceb and sinkb access after buffer_ackquire().